Add pubsys: make Bottlerocket repos using `cargo make repo`

zmrow opened the pull request:
Simplify repo creation and tie it into the existing `cargo make` build system
using a new `repo` target.  This is intended to replace the common pattern of
calls to tuftool and updata - either creating a repo or extending an existing
one, adding the latest built artifacts, then updating the manifest to match.

**Usage:**
`cargo make repo` depends on the `build` target, so if your goal is building an
update for a repo you don't have to separately build it first.  It uses the
same `BUILDSYS_VARIANT` and `BUILDSYS_ARCH` variables to determine the metadata
for the update added to the repo.

**Requirements:**
* An Infra.toml file, based on Infra.toml.example, listing paths to keys,
  existing repos, etc.
* Release.toml updated with a new version

**Optional further configuration:**
* Repo expiration policy - by default, uses a policy file with 2 week target
  and snapshot expiration and 1 week timestamp expiration.
* Wave policy - same policy files you give to updata today; defaults to
  "default" wave policy.
* Release start time - when waves start and when expiration starts counting
  down; defaults to now.
* Can select which named repo and signing key to use from Infra.toml.

**Design decisions:**
* Built repos are written to a directory like
  /repo/bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d that includes variant,
  arch, and full version in the name, so that you can prepare repos for
  multiple releases in parallel if desired.  Inside are target/ and metadata/
  directories that can be synced to your final repo location.
* buildsys uses environment variables set by cargo-make; we opted instead for
  more standard arg parsing.  It seems more likely that someone would use
  pubsys separately from cargo-make, and pubsys has more input information, so
  arg parsing was clearer.
* cargo-make environment variable expansion is done in phases, and you can't
  refer to a variable defined in the same section if you intend to let the user
  override the earlier variable on the command line.  If you do, the variable
  won't expand, as seen in
  https://github.com/bottlerocket-os/bottlerocket/pull/963.  Because of this,
  until we figure out a better strategy, a couple of variables can't be
  overridden - the path to Release.toml (which we made a variable in this
  change) and the repo output directory.

---

**Testing done:**

* Tested arguments to pubsys, including the optional release start time, and
  the --link-target argument not used by Makefile
* Tested overriding all environment variables in Makefile.toml
* Tested with file- and SSM-based keys
* Tested building a new repo and then extending it, and tested starting from
  live Bottlerocket repo
* Tested using the repo we built based on the Bottlerocket repo, adding a
  v0.4.1 update, and using it with an instance to successfully update

```
repo/
└── bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty
    ├── metadata
    │   ├── 1593646164.snapshot.json
    │   ├── 1593646164.targets.json
    │   ├── 1.root.json
    │   └── timestamp.json
    └── targets
        ├── 1c4024695070dcbc4af6660422f08f2b70e04fc309d40da1fe14d8af5935c2e7.bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-root.verity.lz4 -> /home/user/bottlerocket/bottlerocket/build/images/x86_64-aws-k8s-1.15/bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-root.verity.lz4                                      
        ├── 281141246fa284e932c16a015a0420c0d933a517f1ea3ca5cf4cca20a516f592.bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-root.ext4.lz4 -> /home/user/bottlerocket/bottlerocket/build/images/x86_64-aws-k8s-1.15/bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-root.ext4.lz4                                          
        ├── 70d6664cc5d33be9465377404a81ac91123f1aa08e0ea4d282d030482ff1af57.migrate_v0.3.2_admin-container-v0-5-0.lz4
        ├── 7a28a586ba12273372dfbd9d644d6b573c5ba517f10f1a69cae4356b3867872d.manifest.json
        └── 96cf685a8b8292a1a589e9124e0bddbbdad9b12634c252ce257491f078eff257.bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-boot.ext4.lz4 -> /home/user/bottlerocket/bottlerocket/build/images/x86_64-aws-k8s-1.15/bottlerocket-aws-k8s-1.15-x86_64-0.4.1-5880e5d-dirty-boot.ext4.lz4                                          
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
Add pubsys: make Bottlerocket repos using `cargo make repo`

Simplify repo creation and tie it into the existing `cargo make` build system
using a new `repo` target.  This is intended to replace the common pattern of
calls to tuftool and updata - either creating a repo or extending an existing
one, adding the latest built artifacts, then updating the manifest to match.

**Usage:**
`cargo make repo` depends on the `build` target, so if your goal is building an
update for a repo you don't have to separately build it first.  It uses the
same `BUILDSYS_VARIANT` and `BUILDSYS_ARCH` variables to determine the metadata
for the update added to the repo.

**Requirements:**
* An Infra.toml file, based on Infra.toml.example, listing paths to keys,
  existing repos, etc.
* Release.toml updated with a new version

**Optional further configuration:**
* Repo expiration policy - by default, uses a policy file with 2 week target
  and snapshot expiration and 1 week timestamp expiration.
* Wave policy - same policy files you give to updata today; defaults to
  "default" wave policy.
* Release start time - when waves start and when expiration starts counting
  down; defaults to now.
* Can select which named repo and signing key to use from Infra.toml.

**Design decisions:**
* Built repo metadata is written to a directory like
  /build/repos/bottlerocket-0.4.1-5880e5d/aws-k8s-1.15/x86_64 so that you can
  prepare repos for multiple releases in parallel.  Targets are written to
  a shared directory like /build/repos/bottlerocket-0.4.1-5880e5d/targets -
  they're unique across variants and arches so there's no conflict.  The
  directory structure as a whole can be synced to your final repo
  location; it's the structure expected by Bottlerocket and updog.
* buildsys uses environment variables set by cargo-make; we opted instead for
  more standard arg parsing.  It seems more likely that someone would use
  pubsys separately from cargo-make, and pubsys has more input information, so
  arg parsing was clearer.
* cargo-make environment variable expansion is done in phases, and you can't
  refer to a variable defined in the same section if you intend to let the user
  override the earlier variable on the command line.  If you do, the variable
  won't expand, as seen in
  https://github.com/bottlerocket-os/bottlerocket/pull/963.  Because of this,
  until we figure out a better strategy, a couple of variables can't be
  overridden - the path to Release.toml (which we made a variable in this
  change) and the repo output directory.

Co-authored-by: Zac Mrowicki <mrowicki@amazon.com>
Co-authored-by: Tom Kirchner <tjk@amazon.com>

bcressey commented on Makefile.toml:
```diff
@@ -89,7 +112,7 @@ fi
 dependencies = ["setup"]
 script = [
```
add `script_runner = "bash"` since we are using brace expansion

bcressey commented on tools/pubsys/Infra.toml.example:
```diff
@@ -0,0 +1,17 @@
+# Path to your root role JSON file.
+root_role_path = "/home/user/root.json"
+
+# You can keep your private key in a file or an SSM parameter; pick one:
+# (Need inline table syntax until this is fixed: https://github.com/alexcrichton/toml-rs/issues/225)
+signing_keys = { default = { file = { path = "/home/user/key.pem" } } }
+# signing_keys = { default = { ssm = { parameter = "/my/parameter" } } }
```
"private key in a file" is convenient but also feels like a trap for the unwary.

Can we assert that the file is only readable by the user, and not stored under `BUILDSYS_ROOT_DIR`?

tjkirch replied:
Why shouldn't it be under `BUILDSYS_ROOT_DIR`?  What if they're building and using temporary keys for test repos?  I feel like these checks might be assuming too much about workflow..?

tjkirch replied:
I was thinking in too general a sense, I suppose; pubsys isn't for general tough repos that you might use outside of an image context, it's meant for images.  Images would have a hard time using a local repo; it'd have to be a qemu mapped path or something.  I'm fine with more restrictions until we have a reason not to.

tjkirch replied:
After talking about this offline, for now we agreed to gitignore pem files at the root of the repo, to catch some bad usages, and to add a comment in the example Infra.toml saying not to store keys in the repo.  We agreed not to go further than this for now because we plan to automate role/key creation with a make target, and then gitignore the keys directory, or something similar.

bcressey commented on tools/pubsys/Infra.toml.example:
```diff
@@ -0,0 +1,17 @@
+
+[repo.default]
+# metadata_base_url = "https://updates.bottlerocket.aws/2020-02-02/"
+# targets_url = "https://updates.bottlerocket.aws/targets/"
```
I don't think we should point to our URLs here, since they're not going to be useful without our keys.

bcressey commented on tools/pubsys/Infra.toml.example:
```diff
@@ -0,0 +1,17 @@
+# Path to your root role JSON file.
+root_role_path = "/home/user/root.json"
+
+# You can keep your private key in a file or an SSM parameter; pick one:
```
Maybe add a quick example of generating these files.

bcressey commented on Makefile.toml:
```diff
@@ -181,6 +211,7 @@ script = [
 [ "${BUILDSYS_ALLOW_FAILED_LICENSE_CHECK}" = "true" ] && set +e
 (cd sources && cargo deny check --disable-fetch licenses)
 (cd tools/buildsys && cargo deny check --disable-fetch licenses)
+(cd tools/pubsys && cargo deny check --disable-fetch licenses)
```
It might cut down on build times if we combine `buildsys` and `pubsys` in a workspace so they can share dependencies.

bcressey commented on Makefile.toml:
```diff
@@ -234,14 +262,66 @@ alias = "build"
+# Builds a local repository based on the 'latest' built targets.  Uses pubsys
+# to create a repo under /repo, named after the arch/variant/version,
+# containing subdirectories for the repo metadata and targets.
+[tasks.repo]
+dependencies = ["build"]
+script_runner = "bash"
+script = [
+'''
```
This is long enough that it might be better as a helper script (like `tools/docker-go`).

tjkirch replied:
There's a lot of data that pubsys needs from cargo-make and its environment.  docker-go uses arg parsing, but I think that'd be overwhelming and unhelpful with pubsys; would you recommend assuming that environment variables are set?  I worry about that too, because if it's in Makefile.toml directly, at least it's trivial to check if the variables are set.

bcressey commented on Makefile.toml:
```diff
@@ -234,14 +262,66 @@ alias = "build"
+]
+
 [tasks.clean]
 script = [
```
`script_runner = "bash" `

webern commented on tools/pubsys/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+lazy_static = "1.4"
+log = "0.4"
+parse-datetime = { path = "../../sources/parse-datetime" }
+# Need to bring in reqwest with a TLS feature so tough can support TLS repos.
```
Is this an issue with `tough` features that should we can address in `tough`?

```rust
# Need to bring in reqwest with a TLS feature so tough can support TLS repos.
```

tjkirch replied:
We raised this with @iliana and she suggested this workflow, so it's not an issue of features but one of documentation in `tough`, I suppose.

webern commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+fn update_manifest(repo_args: &RepoArgs, manifest: &mut Manifest) -> Result<()> {
+    // Add update   =^..^=   =^..^=   =^..^=   =^..^=
+
+    let filename = |path: &PathBuf| -> Result<String> {
+        Ok(path
+            .file_name()
+            .context(error::InvalidImagePath { path })?
+            .to_str()
+            .context(error::NonUtf8Path { path })?
+            .to_string())
+    };
```
Maybe move it out to a function instead of creating a lambda.

```rust
    let filename = |path: &PathBuf| -> Result<String> {
        Ok(path
            .file_name()
            .context(error::InvalidImagePath { path })?
            .to_str()
            .context(error::NonUtf8Path { path })?
            .to_string())
    };
```

zmrow replied:
We considered that but didn't see us needing it anywhere else but here. If we end up needing it elsewhere we can create a function. :)

webern commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+
+/// Builds an editor and manifest; will start from an existing repo if one is specified in the
+/// configuration
+fn build_editor_and_manifest(
```
The complexity and indenting of this function are fairly high. Consider extracting a function or two. Consider if any subroutines are testable.

```rust
fn build_editor_and_manifest(
```

zmrow replied:
We wrestled with this quite a bit. Unfortunately much of the logic needs to remain together as `tough::Repository` doesn't own much and has lifetimes associated with it (as you've found elsewhere). Quite a bit of this is boilerplate just to load a repo and passing around a `Repository` is difficult.

webern commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+        // Repository is generic over Transport, but we want to take the same actions for file and
+        // HTTP repos.  This function lets us take those common actions regardless of Transport
+        // type, without complicating or repeating the remaining code below.
+        fn editor_and_manifest_from_repo<'a, P: AsRef<Path>, T: Transport>(
```
Is it beneficial for this function to be defined within its calling function like this? Is it so that the generic type does not bubble up to the signature of `build_editor_and_manifest`? I wonder if `tough` `Repository` could take a `Box<dyn Transport>` instead of a generic.

```rust
        fn editor_and_manifest_from_repo<'a, P: AsRef<Path>, T: Transport>(
```

zmrow replied:
The issue stems from the fact that we need to have a loaded `Repository` in order to retrieve the manifest from it. A `Repository<T>` has different types depending on the `Transport` so we didn't want to repeat the manifest retrieval logic in an `if` statement or something. The embedded function allows us to pass any any kind of `Repository<T>` and get what we need from it and centralize any error handling logic as well.

webern reviewed (approved):
Nice! 🍔

etungsten reviewed (approved):
Awesome work!

bcressey commented on tools/pubsys/src/config.rs:
```diff
@@ -0,0 +1,112 @@
+    let path = path.as_ref();
+    let wave_data = fs::read_to_string(path).context(error::File { path })?;
+    toml::from_str(&wave_data).context(error::InvalidToml { path })
+}
```
Should these be associated functions on `Release` and `UpdateWaves`?

bcressey commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+    /// Path to the image containing the verity hashes
+    hash_image: PathBuf,
+
+    // Optionally add other files to the repo
+    #[structopt(long = "link-target", parse(from_os_str))]
```
Do we need this? It seems cleaner to have all repos be independent copies.

tjkirch replied:
I don't think we want to copy all types of targets, particularly large objects like images.  Not everyone has that space to spend, and they might not even realize they're spending it here.

zmrow commented:
Rebased on `develop` and fixed the `Makefile.toml` conflict

bcressey commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+    repo_args: &RepoArgs,
+    infra_config: &InfraConfig,
+) -> Result<(RepositoryEditor, Manifest)> {
+    let repo_config = infra_config
+        .repo
+        .get(&repo_args.repo)
+        .context(error::MissingRepoConfig {
+            repo: &repo_args.repo,
+        })?;
+
+    // If repo section is defined, we have a matching repo name, and it has URLs defined, then try
+    // and load it, otherwise start a new repo
+    // (TODO: use Option.zip when stabilized)
+    let maybe_urls = repo_config.metadata_base_url.as_ref().and_then(|md_url| {
+        repo_config
+            .targets_url
+            .as_ref()
+            .map(|tgt_url| (md_url, tgt_url))
+    });
+    if let Some((metadata_url, targets_url)) = maybe_urls {
+        // Setup to load existing repo   =^..^=   =^..^=   =^..^=   =^..^=
+
+        info!("Found metadata and target URLs, loading existing repository");
+        let base_slash = if metadata_url.as_str().ends_with('/') {
+            ""
+        } else {
+            "/"
+        };
+        let metadata_url_str = format!(
+            "{}{}{}/{}",
+            metadata_url, base_slash, repo_args.variant, repo_args.arch
+        );
+        let metadata_url = Url::parse(&metadata_url_str).context(error::ParseUrl {
+            input: &metadata_url_str,
+        })?;
+        debug!("Using metadata url: {}", metadata_url);
```
It might improve readability to split out all the URL parsing into a helper, and only call this function with valid, formatted URLs. The "no URLs" case is pretty short and could go into `run()` or its own helper.

I share the concern that the complexity of this function is pretty high. I understand what's going on, but it'd be nice to do something else with the nested `editor_and_manifest_from_repo` function. Maybe a macro?

tjkirch replied:
> It might improve readability to split out all the URL parsing into a helper, and only call this function with valid, formatted URLs.

We know the URLs coming from the config are valid by virtue of deserializing into `Url`.  The parsing here is effectively just checking that the variant and arch are valid as URL components, which is true unless the user set the wrong variables.

I'm not sure I understand what you're suggesting to split out.  We could have a function that takes the base URL and components and builds/parses a new one:  `fn metadata_url(metadata_base_url: Url, repo_args: RepoArgs) -> Result<Url>`  Calling that here would split out about 7 lines.  I'm not sure that it'd be worth the separation, but it's just stylistic and we could do that if you think it's clearer.  Or if I'm misinterpreting the request, please let me know.

> The "no URLs" case is pretty short and could go into run() or its own helper.

The goal of this function is abstracting where we get the editor and manifest - putting us in a consistent state.  I think moving the (large) conditional and the "new repo" case into the outer function makes things less consistent, and makes the outer function too large.

> I share the concern that the complexity of this function is pretty high. I understand what's going on, but it'd be nice to do something else with the nested editor_and_manifest_from_repo function. Maybe a macro?

Balancing this code was hard, and obviously not everyone agrees about the best way; there's just a bunch of straight-line actions we have to take.  :/

I think making the inner function a macro wouldn't help with complexity - it'd be basically the same number of lines, but add the complexity of macros, which are used sparingly in Rust.  Moving the inner function out to the top level would mean too many places to look for building an editor/manifest, and make you jump around the code even more when reading.  We tried to break things up into the main logical pieces, and keep each function consistent in achieving its single goal.  We thought hard about it and tried several alternatives and felt this was the best.  We'll continue to think, and welcome suggestions, and agree that this function is one of the more complex ones, but the relatively low-level logic of dealing with tough has to go somewhere, and this felt the best.

tjkirch replied:
We talked about this offline and got a better sense of the organization you were suggesting.  It worked out well and will be in the coming revision.

bcressey commented on tools/pubsys/src/repo.rs:
```diff
@@ -0,0 +1,592 @@
+            source: tough::error::Error,
+        },
+
+        #[snafu(display("Failed to build Target from path '{}': {}", path.display(), source))]
```
nit:
```suggestion
        #[snafu(display("Failed to build target from path '{}': {}", path.display(), source))]
```

tjkirch replied:
The "T" was there to represent it's building a Target object from that path, but I agree with what I think you're getting at, which is that a user shouldn't have to understand that, or even question why it's capitalized.  I'm going to use "target metadata" as a clearer version.

tjkirch commented:
@zmrow and I addressed Ben's comments locally, but while testing them we realized something about the current structure that gave us pause - pubsys can't use one of its output repos as an input for a following run, because of the directory structure.  We need to either decide that it's expected and move on as-is, or come up with a different directory structure that allows for it.

As to why it happens - pubsys writes out to a flat directory structure of `/repo/bottlerocket-VARIANT-ARCH-VERSION/{metadata,targets}`.  This was done so that separate runs have separate results, while keeping them in an obvious location.  However, the expected structure for an input repository (what you point to when you want to add to an existing repo) is more nested, `METADATA_BASE_URL/VARIANT/ARCH` and `TARGETS_BASE_URL`.  This matches the Bottlerocket live repo structure, where you want to share targets rather than duplicate.

One option for a change would be to use the nested structure for pubsys output, too.  The main difference from the current revision is that different pubsys runs would copy/link targets into the same directory, `/repo/bottlerocket-VERSION/targets`.  That's different than our current process, which builds repos in entirely separate directories, but I think it's safe because it'd still be under a `bottlerocket-VERSION` directory (artifacts within a version should be consistent) and variants/arches don't share target names (consistent snapshots cause unique filenames for things like the manifest).

I think the tool can be a little opinionated about structure, rather than needing to have more required variables or a super flexible path system at the moment, so that seems OK to me...

zmrow commented:
I think I'd rather not have a unified `/targets` directory, even under a `repo/bottlerocket-VERSION/targets` folder. It feels a bit confusing to me and I think it could cause some problems and confusion when syncing the repo from your local machine elsewhere.

I'm wondering how often a user would want to use an outdir as the pubsys indir rather than just creating a new repository. Until we have a good answer to that question or an example of a solid use case, my vote is to leave pubsys as we currently have it. I'd rather get some more experience using it and an additional set of use cases before we make changes.

bcressey commented:
> I think I'd rather not have a unified `/targets` directory, even under a `repo/bottlerocket-VERSION/targets` folder. It feels a bit confusing to me and I think it could cause some problems and confusion when syncing the repo from your local machine elsewhere.

Having the property where input, output, and remote structures match seems better to me. I'd like to be able to run `aws s3 sync` from a single directory to update metadata + targets for all variants of interest. It's easier to dry run that way; it's easier to step away for a moment and come back and not remember to do the other half of the sync.

We can split it out into two syncs for production use cases where it's important that targets go up before metadata, but otherwise it just seems more convenient.

We could also make it more convenient to use the repositories by modeling the shared prefix (`https://updates.bottlerocket.aws/2020-02-02/`) as a setting so that it's easy to override just one field in userdata to change both URLs.

tjkirch commented:
We discussed the output repo structure offline, and decided for now to go with the plan to make the output structure match the input structure.  It's clearer to have a single structure, and it's what Bottlerocket is using in production and what we'd recommend to others as well.

zmrow commented:
* Address bcressey's comments above.                                                                                                                                                                                                
* Per above discussion, make output structure under /repo match the input                                                                                                                                                           
  structure, with `/variant/arch` directories for metadata and a shared targets                                                                                                                                                     
  directory.  This allows easily building on top of existing local repos.                                                                                                                                                           
* Before, we would fail if `Infra.toml` contained URLs for the requested repo                                                                                                                                                       
  but `tough` couldn't load a repo from those URLs.  Now, with nested repo                                                                                                                                                          
  paths, the `metadata_base_url` can represent multiple repos, for different                                                                                                                                                        
  variants and arches.  It's reasonable to configure the URLs in `Infra.toml`                                                                                                                                                       
  and add more variants/arches later, so we can't fail if the URLs are already                                                                                                                                                      
  set.  (We still fail if there's something at the URLs that we can't load.)                                                                                                                                                        
* Add a RepoTransport to accomplish the above.                                                                                                                                                                                      
* Simplified editor/manifest construction code; we didn't need the nested                                                                                                                                                           
  function or other complication with generics because of RepoTransport.                                                                                                                                                            
* Now we fail if we load an existing repo and it doesn't have a `manifest.json`.                                                                                                                                                    
* Copy/link targets before writing metadata so we don't have invalid metadata                                                                                                                                                       
  if targets fail                                                                                                                                                                                                                   
* Restructure the new variables we added to more easily accomodate the repo paths.                                                                                                                                                  
* Add `/repo/latest` symlink pointing to the latest built repo, like                                                                                                                                                                
  `/repo/bottlerocket-0.4.0-35e0b5a`.                                                                                                                                                                                               
* Add `.gitignore` entries for the new paths.                                                                                                                                                                                       
                                                                                                                                                                                                                                    
Note: before merging, this requires a new `tough` release to get the changes to                                                                                                                                                     
`copy/link_targets`. We'll need to rebuild `Cargo.lock` in pubsys to match.

zmrow commented:
Rebased and fixed trivial conflicts

bcressey commented on .gitignore:
```diff
@@ -5,3 +5,7 @@
 /.gomodcache
 /html
+/Infra.toml
+/*.pem
```
I'd prefer to have a dedicated directory for keys and exclude everything inside that, especially if we're opinionated and take a path for the key that's assumed to be under that directory.

I'm OK with a follow-up issue for this but would like to capture the long-term goal.

tjkirch replied:
This was the short-term workaround we agreed on until there's a standard process (in Makefile.toml) for making keys that would live in a known path (which would allow simplifying assumptions elsewhere).

tjkirch replied:
https://github.com/bottlerocket-os/bottlerocket/issues/999

bcressey commented on .gitignore:
```diff
@@ -5,3 +5,7 @@
+/Infra.toml
+/*.pem
+/repo
+/tools/bin
```
I'd prefer this to go in `tools/.gitignore` - maybe add `/bin/pubsys` there.

bcressey commented on Makefile.toml:
```diff
@@ -231,14 +267,69 @@ dependencies = [
 done
-rm -f ${BUILDSYS_TOOLS_DIR}/bin/buildsys
+rm -f ${BUILDSYS_TOOLS_DIR}/bin/{buildsys,pubsys}
 for ext in tar lz4 img ; do
   rm -f ${BUILDSYS_OUTPUT_DIR}/*.${ext}
 done
```
Unrelated to this change, but this needs to be fixed now that we no longer store artifacts directly under `${BUILDSYS_OUTPUT_DIR}`.

tjkirch replied:
Similar to https://github.com/bottlerocket-os/bottlerocket/pull/964#discussion_r457693318 - BUILDSYS_OUTPUT_DIR is the full path to arch/variant images, so I believe this command is correct as-is.

bcressey replied:
Correct! I'd mixed up `BUILDSYS_OUTPUT_DIR` and `BUILDSYS_BUILD_DIR`.

bcressey commented on Makefile.toml:
```diff
@@ -87,9 +117,10 @@ fi
+for ws in sources packages variants/* tools/{buildsys,pubsys}; do
   [ -d "${ws}" ] || continue
   cargo fetch --locked --manifest-path ${ws}/Cargo.toml
 done
```
Since they share a workspace now:
```suggestion
for ws in sources packages variants/* tools ; do
  [ -d "${ws}" ] || continue
  cargo fetch --locked --manifest-path ${ws}/Cargo.toml
done
```

bcressey commented on Makefile.toml:
```diff
@@ -10,8 +10,28 @@ BUILDSYS_TOOLS_DIR = "${BUILDSYS_ROOT_DIR}/tools"
+# update waves and repo metadata expiration times will start, instead of
+# starting now.  (This can be an RFC3339 date, or an offset like "in X
+# hours/days/weeks".)
+REPO_EXPIRATION_POLICY_PATH = "${BUILDSYS_ROOT_DIR}/tools/pubsys/policies/repo-expiration/2w-2w-1w.toml"
+REPO_WAVE_POLICY_PATH = "${BUILDSYS_ROOT_DIR}/sources/updater/waves/default-waves.toml"
+REPO_INFRA_CONFIG_PATH = "${BUILDSYS_ROOT_DIR}/Infra.toml"
+# Default repo to read from REPO_INFRA_CONFIG_PATH
+REPO = "default"
+# Default signing key to use from REPO_INFRA_CONFIG_PATH
+KEY = "default"
```
I would rather have all relevant environment variables prefixed with `PUBSYS_` so they're unlikely to be set by default or have other side effects.

`KEY` and `REPO` especially seem very generic.

tjkirch replied:
I'd rather go the other direction and remove the prefix from "BUILDSYS_*" for variables that the user is likely to change, like variant.  That was our methodology here - short variables for things likely to be overridden.  I think it kind of sucks to have to remember these prefixes, especially when you otherwise have no idea what buildsys or pubsys are, since as a user you just deal with "cargo make".

These don't get set by environment variables, only `-e`, so I'm not as concerned about namespacing.  I just tested exporting a bogus BUILDSYS_VARIANT in my shell and it's ignored.

bcressey replied:
The prefixed form is the current convention in the makefile, so I'd like to stick to it.

The prefixes will disambiguate it from other future use cases, so we don't end up with `KEY` (for TUF repos) and `BUILDSYS_KEY` (for signed boot) or similar. They're also easier to grep for when refactoring, and they don't match unpacked source archives that might be present in the tree.

tjkirch replied:
Then I propose that we rename them `BUILD_` and `REPO_` because, again, no one knows (or should have to know) what buildsys and pubsys are.  We could start with `REPO_` in this PR and fix buildsys soon after.

bcressey replied:
`BUILD_` and `PUBLISH_` perhaps? `REPO_REPO` would be odd.

tjkirch replied:
Well, we'd go with something not-odd like `REPO_NAME` instead, but I think `PUBLISH_` is fine because `REPO_` implies the existence of `AMI_`/`SSM_`, but we'd want more generic names for those to cover other platforms, and I can't think of other names atm.

bcressey commented on Makefile.toml:
```diff
@@ -43,6 +63,16 @@ CARGO_MAKE_CARGO_ARGS = "--jobs ${BUILDSYS_JOBS} --offline --locked"
+BUILDSYS_NAME_VERSION="${BUILDSYS_NAME}-${BUILDSYS_VERSION_FULL}"
+BUILDSYS_NAME_FULL="${BUILDSYS_NAME_VARIANT}-${BUILDSYS_VERSION_FULL}"
+# Repo directories have subdirectories for variant/arch, so we only want version here.
+REPO_BASE_DIR = "${BUILDSYS_ROOT_DIR}/repo"
```
Can we move this into `${BUILDSYS_OUTPUT_DIR}` instead?

That way there's a single obvious directory to clean up, and to manage exclusions for.

Otherwise we to remember to duplicate the rules in .dockerignore to also cover images written to repo, or incremental builds will get much slower.

We've already had several regressions like this even with only the one directory to worry about.

tjkirch replied:
Per offline discussion, we want BUILDSYS_BUILD_DIR here; BUILDSYS_OUTPUT_DIR is /build/images/ARCH-VARIANT and repos share targets for arches/variants.

tjkirch commented:
^ This force-push was just a rebase with no conflicts; fixes for comments are coming soon.

tjkirch commented:
This push addresses most of @bcressey's comments:
* Move gitignore of `tools/bin/pubsys` to `tools/.gitignore`
* Move `/repo` to `/build/repos`
* Remove gitignore of `/repo` since it's covered by existing gitignore of `/build` now
* Only fetch once for `tools` workspace
* Use relative link for `latest` to make clear it's in the same directory
* Add `clean` step for repos
* Fix variant names for `PathExists` to match update in https://github.com/awslabs/tough/pull/187/

tjkirch commented:
^ This push updates variable names in `Makefile.toml` with a `PUBLISH_` prefix, per above discussion with @bcressey.

zmrow commented:
This force-push updates the Cargo.lock to reflect the new `tough` version.
